### PR TITLE
2.4 into 2.5

### PR DESCRIPTION
**Pull request by jameinel**
## Description of change

Merge of the current tip of 2.4 into 2.5 bringing in patches:
 * Merge pull request #9359 from juju/2.3
 * Merge pull request #9355 from howbazaar/2.4-engine-backoff
 * Merge pull request #9345 from anastasiamac/only-clear-current-when-asked-lp1796148
 * Merge pull request #9335 from babbageclunk/mux-worker
 * Merge pull request #9333 from wallyworld/merge-2.3-20181018
 * Merge pull request #9323 from ExternalReality/fix_logging_path
 * Merge pull request #9327 from mitechie/enable-export-bundle
 * Merge pull request #9315 from manadart/2.4-acceptance-maas-network-bind
 * Merge pull request #9300 from manadart/2.4-provisioner-testing-with-mocks
 * Merge pull request #9287 from manadart/2.4
 * Merge pull request #9297 from anastasiamac/24-better-debug-lp1779677
 * Merge pull request #9298 from wallyworld/cmr-controller-connect-robust-2.4
 * Merge pull request #9251 from howbazaar/bundle-export-series
 * Merge pull request #9293 from babbageclunk/bundle-deploy-panic-2.4
 * Merge pull request #9285 from howbazaar/2.4-fix-multiwatcher
 * Merge pull request #9264 from mitechie/forward-9235
 * Merge pull request #9263 from wallyworld/goal-state-units
 * Merge pull request #9255 from howbazaar/2.4-bionic-tests
 * Merge pull request #9203 from wallyworld/go-1.11-2.4
 * Merge pull request #9191 from howbazaar/2.4-status-retry
 * Merge pull request #9178 from hmlanigan/fixes
 * Merge pull request #9177 from SimonRichardson/ec2-instance-types-2.4
 * Merge pull request #9181 from babbageclunk/2.3-merge-into-2.4
 * Merge pull request #9169 from juju/2.3
 * Merge pull request #9167 from babbageclunk/caasfirewaller-test-fix-2.4
 * Merge pull request #9166 from babbageclunk/s390x-test-fix-2.4
 * Merge pull request #9152 from jameinel/2.3-into-2.4
 * Merge pull request #9080 from vinu2003/juju_exportBundle_timeDelayFix
 * Merge pull request #9138 from wallyworld/charm.v6-dep-update
 * Merge pull request #9137 from anastasiamac/test-order-fix-24
 * Merge pull request #9134 from anastasiamac/merge-23-into-24-20180829
 * Merge pull request #9132 from wallyworld/improve-charm-versionstring
 * Merge pull request #9126 from anastasiamac/empty-default-settings-lp1789415-24
 * Merge pull request #9117 from anastasiamac/init-cmd-lp1785205-24
 * Merge pull request #9104 from wallyworld/merge-2.3-20180823

## QA steps

See individual patches.

## Documentation changes

See individual patches.

## Bug reference

 * https://bugs.launchpad.net/juju/+bug/1798485
 * https://bugs.launchpad.net/juju/+bug/1793245
 * https://bugs.launchpad.net/juju/+bug/1796148
 * https://bugs.launchpad.net/juju/+bug/1798001
 * https://bugs.launchpad.net/juju/+bug/1793284
 * https://bugs.launchpad.net/juju/+bug/1796106
 * https://bugs.launchpad.net/juju/+bug/1779677
 * https://bugs.launchpad.net/juju/+bug/1795499
 * https://bugs.launchpad.net/juju/+bug/1773357
 * https://bugs.launchpad.net/juju/+bug/1792299
 * https://bugs.launchpad.net/juju/+bug/1782803
 * https://bugs.launchpad.net/juju/+bug/1782367
 * https://bugs.launchpad.net/juju/+bug/1790626
 * https://bugs.launchpad.net/juju/+bug/1789211
 * https://bugs.launchpad.net/juju/+bug/1789447
 * https://bugs.launchpad.net/juju/+bug/1789415
 * https://bugs.launchpad.net/juju/+bug/1785205
 * https://bugs.launchpad.net/juju/2.3/+bug/1788554

### CODE REVIEW

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
self approving a merge of 2.4 into 2.5

File: worker/apiserver/worker_state_test.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ var _ = gc.Suite(&WorkerStateSuite{})
 func (s *WorkerStateSuite) SetUpSuite(c *gc.C) {
 	s.workerFixture.SetUpSuite(c)
 
-	err := testing.MgoServer.Start(nil)
+	err = testing.MgoServer.Start(nil)
```
Comment (wallyworld) [inline worker/apiserver/worker_state_test.go]: this won't compile, need to revert the change

File: worker/httpserver/state_test.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ var _ = gc.Suite(&stateFixture{})
 func (s *stateFixture) SetUpSuite(c *gc.C) {
 	s.IsolationSuite.SetUpSuite(c)
 
-	err := testing.MgoServer.Start(nil)
+	err = testing.MgoServer.Start(nil)
```
Comment (wallyworld) [inline worker/httpserver/state_test.go]: ditto

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
$$merge$$ (fixed a couple of other conflicts between how 2.4 and 2.5 were doing things)
